impl/src/main/java/com/okta/sdk/impl/util/OAuth2Utils.java

```diff
@@ -0,0 +1,163 @@
+package com.okta.sdk.impl.util;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.reflect.TypeToken;
+import com.okta.commons.http.*;
+import com.okta.commons.http.okhttp.OkHttpRequestExecutorFactory;
+import com.okta.sdk.impl.config.ClientConfiguration;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import org.bouncycastle.util.io.pem.PemObject;
+import org.bouncycastle.util.io.pem.PemReader;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
+
+public class OAuth2Utils {
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+    private static List<String> allowedSigningAlgorithms =
+        new ArrayList<>(Arrays.asList("RS256", "RS384", "RS512", "ES256", "ES384", "ES512"));
+
+    private static final OkHttpRequestExecutorFactory okHttpRequestExecutorFactory =
+        new OkHttpRequestExecutorFactory();
+
+    private static KeyFactory getKeyFactoryInstance() throws NoSuchAlgorithmException {
+        return KeyFactory.getInstance("RSA");
+    }
+
+    public static String createSignedJWT(ClientConfiguration clientConfiguration)
+        throws RuntimeException, InvalidKeySpecException, NoSuchAlgorithmException, IOException {
+        String clientId = clientConfiguration.getClientId();
+        SignatureAlgorithm signatureAlgorithm = getSignatureAlgorithm(clientConfiguration.getAlgorithm());
+        PrivateKey privateKey = createPrivateKeyFromPemFilePath(clientConfiguration.getPrivateKey());
+        Instant now = Instant.now();
+
+        String jwt = Jwts.builder()
+            .setAudience(clientConfiguration.getBaseUrl() + "/oauth2/default/v1/token")
+            .setIssuedAt(Date.from(now))
+            .setExpiration(Date.from(now.plus(1L, ChronoUnit.HOURS)))
+            .setIssuer(clientId)
+            .setSubject(clientId)
+            .claim("jti", UUID.randomUUID().toString())
+            .signWith(signatureAlgorithm, privateKey)
+            .compact();
+
+        return jwt;
+    }
+
+    public static String getOAuth2AccessToken(ClientConfiguration clientConfiguration)
+        throws HttpException, IOException, InvalidKeySpecException, NoSuchAlgorithmException {
+        String signedJwt = createSignedJWT(clientConfiguration);
+
+        RequestExecutor requestExecutor = okHttpRequestExecutorFactory.create(clientConfiguration);
```

Comment:
This class shouldn't be static, it should be implemented as a service and reused (when possible)
Instead of using the RequestExecutor, create a new instance of a `Client` and make the request all of this logic can be moved there and the JSON will be parsed automatically for you (removing your dependency on Gson)

Reply:
Updated code.